### PR TITLE
Remove usage of componentWillReceiveProps

### PR DESCRIPTION
**Pull request by pradeepnschrodinger**
## Description

Remove the last usage of the `componentWillReceiveProps` lifecycle method in our codebase.

The usage is in `FixedDataTableContainer.js` where the redux store is made to update against the new props passed by the user.
I replaced the lifecycle method with the new static `getDerivedStateFromProps` method.

## Motivation and Context

* [SS-31705](https://jira.schrodinger.com/browse/SS-31705): FDT: Replace deprecated React life cycle methods
* #492

## How Has This Been Tested?

Tested examples and in LD, and everything works fine.
Also verified that there's no additional rerenders happening.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.

### CODE REVIEW

File: src/FixedDataTableContainer.js [hunk 1]
```diff
@@ -45,21 +44,39 @@ class FixedDataTableContainer extends React.Component {
       props,
     });
 
-    this.unsubscribe = this.reduxStore.subscribe(this.update);
-    this.state = this.getBoundState();
+    this.unsubscribe = this.reduxStore.subscribe(this.onStoreUpdate.bind(this));
```
Comment (deshpsuy) [inline src/FixedDataTableContainer.js:47]: Let's make `onStoreUpdate` fat arrow method and remove this bind?

**Review by deshpsuy — APPROVED** [review]

**Comment by tstirrat15** [comment]
What's the status on this one? I'd be keen to see it get through and am happy to help if there's a way to do so.

**Comment by pradeepnschrodinger** [comment]
@miskreant , @novakps , @mycrobe , what do you guys think about the usage of `getDerivedStateFromProps` as an alternative to `componentWillReceiveProps` inorder to update the store based on incoming props?

**Comment by tstirrat15** [comment]
Looking at the code: I think `gDSP` should be similar enough to `cWU` for this to work, but the code generally makes me kind of uneasy. I don't know enough of the context or history for why things were done this way, but sending entire props objects into redux is generally discouraged, and I'd be concerned about the possibility for infinite loops under the right circumstances. `gDSP` is going to run before each render, and if it's triggering updates in the redux store that is going to subsequently trigger another render.

I also am unsure of why this component uses redux internally (as opposed to local state and/or a `useReducer`), and why redux and local state are mixed in the way they are.

Again, this is without me completely understanding why things were done the way they were done, so if there are reasons feel free to discard this and/or take with a grain of salt or three.

**Comment by pradeepnschrodinger** [comment]
@tstirrat15 , thanks those are fair points!

The main reason behind having `cWU` setup like this was to update our redux store in response to incoming props.
Since `cWU` is getting deprecated, we wanted to figure out an alternative approach.

One approach we initially explored was making use of `componentDidUpdate` to dispatch an action to update the redux store.
This had two drawbacks though:
- There's going to be two renders for every corresponding prop change. This got slightly inefficient during controlled scrolling.
- This caused breakage in one of our applications that heavily relies on FDT. The first render passed data based on the older value of the store (because componentDidUpdate hasn't been fired yet), but the cell renderers expect to get props based on the newer value of the store, which only comes in during the second render. There's more information about this on #523.

We then explored `gDSP` which is the focus of this PR.

To answer some of your questions:
> sending entire props objects into redux is generally discouraged

Agreed. 
Currently, the `PROPS_CHANGE` reducer works by comparing the previous and next set of props, so it requires these in the action payload.
However, `gDSP` which triggers the action doesn't have a direct way of accessing the older props object. 
So what we did here was to store the older props in the redux store.
I now realize that a _slightly_ cleaner way would have been to simply store the older props in local state itself instead of in the redux store. I'll follow up with a quick commit to fix this.

> I'd be concerned about the possibility for infinite loops under the right circumstances. `gDSP` is going to run before each render, and if it's triggering updates in the redux store that is going to subsequently trigger another render.

Good point. The potentially cyclic trigger from `gDSP` to the redux store which then triggers the subscriber does happen here. To avoid this, we follow this rule:
If the subscriber was called by `gDSP` then we [ignore the update](https://github.com/schrodinger/fixed-data-table-2/pull/612/files#diff-5b6dc707d07f71daeba791f1764520ee17b03d5186a0af5cc9a692a97ae9419aR148).
Ignoring this update is fine because `gDSP` already took care of updating the local state as soon as it returns the new state object.

**Comment by tstirrat15** [comment]
Sounds good to me, then :+1:

**Comment by pradeepnschrodinger** [comment]
The branch was a bit outdated, so there were some small conflicts related to formatting.

So I merged this offline through febc9f278163b9071c4d37a1e36c263b17bbbad4.
Released with [v1.2.5](https://github.com/schrodinger/fixed-data-table-2/releases/tag/v1.2.5).

**Comment by pradeepnschrodinger** [comment]
I found out that there's a redundant `PROPS_CHANGE` action being dispatched in `getDerivedStateFromProps` when local state is updated, due to a typo when accessing the props object.

Fixed it through [v1.2.6](https://github.com/schrodinger/fixed-data-table-2/releases/tag/v1.2.6).

**Pull request closed without merging**